File: app/src/main/cpp/sources/pbr/mario/camera.cpp
#include <pbr/mario/camera.h>


namespace pbr::mario {

constexpr static float const FIELD_OF_VIEW = 60.0F;
constexpr static float const Z_NEAR = 0.1F;
constexpr static float const Z_FAR = 1.0e+4F;
constexpr static GXVec3 const OFFSET ( 512.0F, 128.0F, 0.0F );
constexpr static float const TOLERANCE = 1.0e-4F;
constexpr static float const SPEED = 2.0e-2F;

Camera::Camera () noexcept:
    _local {},
    _projection {},
    _target ( nullptr )
{
    _local.RotationY ( -GX_MATH_HALF_PI );
    _projection.Identity ();
}

void Camera::Focus () noexcept
{
    if ( !_target )
        return;

    GXVec3 target {};
    _target->GetTransform ().GetW ( target );

    GXVec3 ideal {};
    ideal.Sum ( target, OFFSET );
    _local.SetW ( ideal );
}

GXMat4 const& Camera::GetLocalMatrix () const noexcept
{
    return _local;
}

GXMat4 const& Camera::GetProjectionMatrix () const noexcept
{
    return _projection;
}

void Camera::OnUpdate ( float deltaTime ) noexcept
{
    if ( !_target )
        return;

    GXVec3 target {};
    _target->GetTransform ().GetW ( target );

    GXVec3 ideal {};
    ideal.Sum ( target, OFFSET );

    GXVec3 loc {};
    _local.GetW ( loc );

    GXVec3 delta {};
    delta.Subtract ( ideal, loc );

    float const length = delta.Length ();

    if ( length < TOLERANCE )
        return;

    loc.Sum ( loc, length * SPEED * deltaTime, delta );
    _local.SetW ( loc );
}

void Camera::OnResolutionChanged ( VkExtent2D const &resolution ) noexcept
{
    _projection.Perspective ( GXDegToRad ( FIELD_OF_VIEW ),
        static_cast<float> ( resolution.width ) / static_cast<float> ( resolution.height ),
        Z_NEAR,
        Z_FAR
    );
}

void Camera::SetTarget ( ITarget const &target ) noexcept
{
    _target = &target;
}

} // namespace pbr::mario

File: app/src/main/assets/pbr/assets/Scripts/camera.lua
require "av://engine/script_component.lua"


-- Constants
local FOVY = math.rad ( 60.0 )
local ZNEAR = 1.0e-1
local ZFAR = 1.0e+4

local OFFSET = GXVec3 ()
OFFSET:Init ( 512.0, 128.0, 0.0 )

local SPEED = 2.0e-2
local TOLERANCE = 1.0e-4

-- Class declaration
local Camera = {}

-- Engine event handlers
local function OnActorConstructed ( self, actor )
    local camera = actor:FindComponent ( "Camera" )
    camera:SetProjection ( FOVY, g_scene:GetRenderTargetAspectRatio (), ZNEAR, ZFAR )
    self._cameraComponent = camera
    g_scene:SetActiveCamera ( camera )
end

local function OnRenderTargetChanged ( self )
    self._cameraComponent:SetProjection ( FOVY, g_scene:GetRenderTargetAspectRatio (), ZNEAR, ZFAR )
end

local function Tracking ( self, deltaTime )
    local target = GXVec3 ()
    self._tracker:GetLocation ( target )

    local ideal = GXVec3 ()
    ideal:SumScaled ( OFFSET, g_scene:GetPhysicsToRendererScaleFactor (), target )

    local localMatrix = self._localMatrix
    local loc = GXVec3 ()
    localMatrix:GetW ( loc )

    local delta = GXVec3 ()
    delta:Subtract ( ideal, loc )

    local length = delta:Length ()

    if length < TOLERANCE then
        return
    end

    loc:SumScaled ( loc, length * SPEED * deltaTime, delta )
    localMatrix:SetW ( loc )

    self._cameraComponent:SetLocal ( localMatrix )
end

local function FindTracker ( self, deltaTime )
    local mario = g_scene:FindActor ( "Mario" )

    if not mario then
        return
    end

    local tracker = mario:FindComponent ( "Collider" )

    local target = GXVec3 ()
    tracker:GetLocation ( target )
    target:MultiplyScalar ( target, g_scene:GetPhysicsToRendererScaleFactor () )
    self._localMatrix:SetW ( target )

    self._tracker = tracker

    -- Switching from finding state to tracking state.
    self.OnPostPhysics = Tracking
end

-- Metamethods
local function Constructor ( self, handle, params )
    local obj = ScriptComponent ( handle )

    -- Data
    obj._localMatrix = GXMat4 ()
    obj._localMatrix:RotationY ( math.rad ( -90.0 ) )

    -- Engine events
    obj.OnActorConstructed = OnActorConstructed
    obj.OnPostPhysics = FindTracker
    obj.OnRenderTargetChanged = OnRenderTargetChanged

    return obj
end

setmetatable ( Camera, { __call = Constructor } )

-- Module function: fabric callable for Camera class.
return Camera
